core/src/Cuda/Kokkos_Cuda_Parallel_MDRange.hpp

```diff
@@ -234,10 +241,18 @@ class ParallelReduce<FunctorType, Kokkos::MDRangePolicy<Traits...>, ReducerType,
         CudaParallelLaunch<ParallelReduce,
                            LaunchBounds>::get_cuda_func_attributes();
     auto const& prop = pol.space().cuda_device_prop();
-    // Limits due do registers/SM
-    int const regs_per_sm        = prop.regsPerMultiprocessor;
-    int const regs_per_thread    = attr.numRegs;
-    int const max_threads_per_sm = regs_per_sm / regs_per_thread;
+    // Limits due to registers/SM, MDRange doesn't have
+    // shared memory constraints
+    int const optimal_block_size = Kokkos::Impl::cuda_get_opt_block_size_no_shmem(attr, LaunchBounds{});
+
+    // Compute how many blocks of this size we can launch, based on warp constraints
+    int const max_warps_per_sm_registers = Kokkos::Impl::cuda_max_warps_per_sm_registers(prop, attr);
+    int const max_num_threads_from_warps = max_warps_per_sm_registers * prop.warpSize;
+    int const max_num_blocks = max_num_threads_from_warps / optimal_block_size;
+
+    // Compute the total number of threads
+    int const max_threads_per_sm = optimal_block_size * max_num_blocks;
```

Comment:
This code is the same as above right? Is it possible to put it in a free function to avoid duplication.

Reply:
Yes, it's identical. Were you thinking in this file, or in `Kokkos_Cuda_BlockSize_Deduction.hpp`?

Reply:
Either way. I don't have a preference

Reply:
https://github.com/kokkos/kokkos/pull/5739/commits/3a23689ba3e861d261b7e251eaa56aaa4e906a1f

Reply:
Whoops, left some debugging stuff in there...

Reply:
https://github.com/kokkos/kokkos/pull/5739/commits/dd72c16017f4e02770af40546b4064ae39205716